File: sampling/py/SimpleRandomSample.py
import heapq
import random

# Simple Random Sample: get a sample of size k from input of size n

# Simple Random Sample algorithm for fixed length input
# Intuition: randomly taking 3 balls out of the box at once is equivalent to
#            randomly taking 1 ball from the box without replacement 3 times
# Time:  O(n)
# Space: O(k)
def selection_rejection(array, k):
    """ Fan et al., 1962 """
    sample = []
    n = len(array)
    for elem in array:
        if random.random() * n < k:
            sample.append(elem)
            k -= 1
        n -= 1
    return sample

# Simple Random Sample algorithm for stream input
# Every incoming element has a k/n chance of being sampled
# Time:  O(n)
# Space: O(k)
def reservoir_sampling(iterator, k):
    """ Vitter, 1985 """
    sample = []
    for i in range(k):
        try:
            sample.append(next(iterator))
        except StopIteration:
            return sample

    n = k
    while True:
        n += 1
        try:
            elem = next(iterator)
            if random.random() * n < k:
                idx = random.randrange(k)
                sample[idx] = elem
        except StopIteration:
            break
    return sample

File: sampling/cxx/simple_random_sample.cxx
#include <algorithm>
#include <cstdlib>
#include <vector>

template<class T>
std::vector<T> draw_by_draw(std::vector<T>& array, int k)
{
	int index;
	for(int i = 0; i < k; i++)
	{
		index = rand() % (array.size() - i);
		std::swap(array[index], array[array.size() - 1 - i]);
	}
	return std::vector<int>(array.begin() + array.size() - k, array.end());
}

template<class T>
std::vector<T> selection_rejection(const std::vector<T>& array, int k)
{
	std::vector<T> sample(k, 0);
	int cnt = 0, n = array.size();
	for(const auto& elem: array)
	{
		if((float)rand() / RAND_MAX < (float)k / n)
		{
			sample[cnt++] = elem;
			k--;
		}
		n--;
	}
	return sample;
}

template<class T>
std::vector<T> reservoir_sampling(const std::vector<T>& array, int k)
{
	std::vector<T> sample(k, 0);
	auto it = array.begin();
	for(int i = 0; i < k; i++)
	{
		sample[i] = *it++;
		if(it == array.end())
			return sample;
	}
	int n = k;
	int idx;
	while(it != array.end())
	{
		n += 1;
		if((float)rand() / RAND_MAX < (float)k / n)
		{
			idx = rand() % k;
			sample[idx] = *it;
		}
		it++;
	}
	return sample;
}
